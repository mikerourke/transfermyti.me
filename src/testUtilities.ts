export { FAKES } from "~/redux/__fakes__/state";

type MockInstance = jest.MockInstance<AnyValid, AnyValid>;

interface WrapForMockingReturn {
  calls: AnyValid[];
  clear(): MockInstance;
  implementOnce(funcOrReturn: AnyValid): MockInstance;
  rejectAlways(err?: AnyValid): MockInstance;
  rejectOnce(err?: AnyValid): MockInstance;
  resolveMultiple(values: AnyValid[]): void;
  resolveOnce(value?: AnyValid): MockInstance;
  returnOnce(value?: AnyValid): MockInstance;
  returnMultiple(values: AnyValid[]): void;
  throwOnce(message?: string): MockInstance;
}

/**
 * Wrappers around someFunc.mock...() that Jest provides. Using
 * this avoids the annoying `someFunc as AnyValid` required to shut TypeScript
 * up.
 */
export function wrapForMocking(func: AnyValid): WrapForMockingReturn {
  return {
    calls: func?.mock?.calls,
    clear(): MockInstance {
      return func.mockClear();
    },
    implementOnce(funcOrReturn: AnyValid): MockInstance {
      if (typeof funcOrReturn === "function") {
        return func.mockImplementationOnce(funcOrReturn);
      } else {
        return func.mockImplementationOnce(() => funcOrReturn);
      }
    },
    rejectAlways(err: AnyValid = "error"): MockInstance {
      return func.mockRejectedValue(
        typeof err === "string" ? new Error(err) : err,
      );
    },
    rejectOnce(err: AnyValid = "error"): MockInstance {
      return func.mockRejectedValueOnce(
        typeof err === "string" ? new Error(err) : err,
      );
    },
    resolveMultiple(values: AnyValid[]): void {
      for (const value of values) {
        func.mockResolvedValueOnce(value);
      }
    },
    resolveOnce(value: AnyValid = null): MockInstance {
      return func.mockResolvedValueOnce(value);
    },
    returnOnce(value: AnyValid = null): MockInstance {
      return func.mockReturnValueOnce(value);
    },
    returnMultiple(values: AnyValid[]): void {
      for (const value of values) {
        func.mockReturnValueOnce(value);
      }
    },
    throwOnce(message: string = "error"): MockInstance {
      return func.mockImplementationOnce(() => {
        throw new Error(message);
      });
    },
  };
}

/**
 * Allows you to get around `delay()` calls in sagas.
 * @see https://github.com/jfairbank/redux-saga-test-plan/issues/257#issuecomment-503016758
 * @example
 *   return expectSaga(someSaga)
 *     .provide([
 *       { call: provideDelay },
 *     ])
 *     ...
 */
export function provideDelay(
  { fn }: AnyValid,
  next: AnyValid,
): null | VoidFunction {
  return fn.name === "delayP" ? null : next();
}
