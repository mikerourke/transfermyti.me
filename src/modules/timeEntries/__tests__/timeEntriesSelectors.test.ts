import { describe, expect, test } from "vitest";

import { FAKES } from "~/testUtilities";

import * as timeEntriesSelectors from "../timeEntriesSelectors";

const MOCK_STATE = {
  ...FAKES.REDUX_STATE,
  timeEntries: {
    ...FAKES.REDUX_STATE.timeEntries,
    source: {
      "8001": {
        ...FAKES.REDUX_STATE.timeEntries.source["8001"],
        entryCount: 0,
        linkedId: "clock-entry-01",
        isIncluded: false,
      },
      "8002": {
        ...FAKES.REDUX_STATE.timeEntries.source["8002"],
        entryCount: 0,
        linkedId: "clock-entry-02",
        isIncluded: false,
      },
      "8003": {
        ...FAKES.REDUX_STATE.timeEntries.source["8003"],
        entryCount: 0,
        linkedId: "clock-entry-03",
        isIncluded: false,
      },
      "8004": {
        ...FAKES.REDUX_STATE.timeEntries.source["8004"],
        entryCount: 0,
        linkedId: null,
        isIncluded: true,
      },
      "8005": {
        ...FAKES.REDUX_STATE.timeEntries.source["8005"],
        entryCount: 0,
        linkedId: null,
        isIncluded: true,
      },
    },
    target: {
      "clock-entry-01": {
        ...FAKES.REDUX_STATE.timeEntries.target["clock-entry-01"],
        entryCount: 0,
        linkedId: "8001",
        isIncluded: true,
      },
      "clock-entry-02": {
        ...FAKES.REDUX_STATE.timeEntries.target["clock-entry-02"],
        entryCount: 0,
        linkedId: "8002",
        isIncluded: true,
      },
      "clock-entry-03": {
        ...FAKES.REDUX_STATE.timeEntries.target["clock-entry-03"],
        entryCount: 0,
        linkedId: "8003",
        isIncluded: true,
      },
    },
  },
};

describe("within timeEntriesSelectors", () => {
  test("the sourceTimeEntriesSelector returns array of values from state.source", () => {
    const result = timeEntriesSelectors.sourceTimeEntriesSelector(MOCK_STATE);

    expect(result).toEqual(Object.values(MOCK_STATE.timeEntries.source));
  });

  describe("the selectors match their snapshots", () => {
    const testCases = [
      {
        name: "for the includedSourceTimeEntriesSelector",
        selector: timeEntriesSelectors.includedSourceTimeEntriesSelector,
      },
      {
        name: "for the sourceTimeEntriesForTransferSelector",
        selector: timeEntriesSelectors.sourceTimeEntriesForTransferSelector,
      },
      {
        name: "for the sourceTimeEntriesInActiveWorkspaceSelector",
        selector: timeEntriesSelectors.sourceTimeEntriesInActiveWorkspaceSelector,
      },
      {
        name: "for the timeEntriesTotalCountsByTypeSelector",
        selector: timeEntriesSelectors.timeEntriesTotalCountsByTypeSelector,
      },
      {
        name: "for the sourceTimeEntryCountByTagIdSelector",
        selector: timeEntriesSelectors.sourceTimeEntryCountByTagIdSelector,
      },
    ];

    for (const testCase of testCases) {
      test.concurrent(testCase.name, async () => {
        const result = testCase.selector(MOCK_STATE);

        expect(result).toMatchSnapshot();
      });
    }
  });

  test(`the sourceTimeEntriesForTransferSelector returns all values if state.isDuplicateCheckEnabled = false`, () => {
    const updatedState = {
      ...MOCK_STATE,
      timeEntries: {
        ...MOCK_STATE.timeEntries,
        isDuplicateCheckEnabled: false,
      },
    };

    const result = timeEntriesSelectors.sourceTimeEntriesForTransferSelector(updatedState);

    const expected = timeEntriesSelectors.includedSourceTimeEntriesSelector(MOCK_STATE);

    expect(result).toEqual(expected);
  });

  describe("the timeEntriesForInclusionsTableSelector matches its snapshot based on state.allEntities.areExistsInTargetShown", () => {
    const testCases = [
      {
        name: "when state.allEntities.areExistsInTargetShown = true",
        areExistsInTargetShown: true,
      },
      {
        name: "when state.allEntities.areExistsInTargetShown = false",
        areExistsInTargetShown: false,
      },
    ];

    for (const testCase of testCases) {
      test.concurrent(testCase.name, async () => {
        const updatedState = {
          ...MOCK_STATE,
          allEntities: {
            ...MOCK_STATE.allEntities,
            areExistsInTargetShown: testCase.areExistsInTargetShown,
          },
        };

        const result = timeEntriesSelectors.timeEntriesForInclusionsTableSelector(updatedState);

        expect(result).toMatchSnapshot();
      });
    }
  });

  test(`the timeEntriesForInclusionsTableSelector matches its snapshot when state.isDuplicateCheckEnabled = false`, () => {
    const updatedState = {
      ...MOCK_STATE,
      timeEntries: {
        ...MOCK_STATE.timeEntries,
        isDuplicateCheckEnabled: false,
      },
    };

    const result = timeEntriesSelectors.timeEntriesForInclusionsTableSelector(updatedState);

    expect(result).toMatchSnapshot();
  });

  test("the sourceTimeEntryCountByIdFieldSelectorFactory matches its snapshot", () => {
    const getCounts = timeEntriesSelectors.sourceTimeEntryCountByIdFieldSelectorFactory("6001");

    const result = getCounts(MOCK_STATE);

    expect(result).toMatchSnapshot();
  });
});
