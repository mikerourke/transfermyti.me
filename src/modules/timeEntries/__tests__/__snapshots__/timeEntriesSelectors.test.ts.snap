// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`within timeEntriesSelectors the selectors match their snapshots for the includedSourceTimeEntriesSelector 1`] = `
[
  {
    "clientId": "3002",
    "description": "This is a test entry (04)",
    "end": 2019-06-23T17:05:20.000Z,
    "entryCount": 0,
    "id": "8004",
    "isActive": false,
    "isBillable": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "timeEntries",
    "projectId": "2002",
    "start": 2019-06-23T14:00:20.000Z,
    "tagIds": [],
    "tagNames": [],
    "taskId": null,
    "userGroupIds": [],
    "userId": "6001",
    "workspaceId": "1001",
    "year": 2019,
  },
  {
    "clientId": "3002",
    "description": "This is a test entry (05)",
    "end": 2019-06-22T19:15:19.000Z,
    "entryCount": 0,
    "id": "8005",
    "isActive": false,
    "isBillable": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "timeEntries",
    "projectId": "2002",
    "start": 2019-06-22T14:00:19.000Z,
    "tagIds": [],
    "tagNames": [],
    "taskId": null,
    "userGroupIds": [],
    "userId": "6001",
    "workspaceId": "1001",
    "year": 2019,
  },
]
`;

exports[`within timeEntriesSelectors the selectors match their snapshots for the sourceTimeEntriesForTransferSelector 1`] = `
[
  {
    "clientId": "3002",
    "description": "This is a test entry (04)",
    "end": 2019-06-23T17:05:20.000Z,
    "entryCount": 0,
    "id": "8004",
    "isActive": false,
    "isBillable": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "timeEntries",
    "projectId": "2002",
    "start": 2019-06-23T14:00:20.000Z,
    "tagIds": [],
    "tagNames": [],
    "taskId": null,
    "userGroupIds": [],
    "userId": "6001",
    "workspaceId": "1001",
    "year": 2019,
  },
  {
    "clientId": "3002",
    "description": "This is a test entry (05)",
    "end": 2019-06-22T19:15:19.000Z,
    "entryCount": 0,
    "id": "8005",
    "isActive": false,
    "isBillable": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "timeEntries",
    "projectId": "2002",
    "start": 2019-06-22T14:00:19.000Z,
    "tagIds": [],
    "tagNames": [],
    "taskId": null,
    "userGroupIds": [],
    "userId": "6001",
    "workspaceId": "1001",
    "year": 2019,
  },
]
`;

exports[`within timeEntriesSelectors the selectors match their snapshots for the sourceTimeEntriesInActiveWorkspaceSelector 1`] = `
[
  {
    "clientId": "3001",
    "description": "This is a test entry (01)",
    "end": 2019-06-25T23:00:00.000Z,
    "entryCount": 0,
    "id": "8001",
    "isActive": false,
    "isBillable": false,
    "isIncluded": false,
    "linkedId": "clock-entry-01",
    "memberOf": "timeEntries",
    "projectId": "2001",
    "start": 2019-06-25T18:00:00.000Z,
    "tagIds": [
      "4001",
      "4002",
    ],
    "tagNames": [
      "tag-a",
      "tag-b",
    ],
    "taskId": "7001",
    "userGroupIds": [],
    "userId": "6001",
    "workspaceId": "1001",
    "year": 2019,
  },
  {
    "clientId": "3001",
    "description": "This is a test entry (02)",
    "end": 2019-06-24T22:00:00.000Z,
    "entryCount": 0,
    "id": "8002",
    "isActive": false,
    "isBillable": false,
    "isIncluded": false,
    "linkedId": "clock-entry-02",
    "memberOf": "timeEntries",
    "projectId": "2001",
    "start": 2019-06-24T14:00:00.000Z,
    "tagIds": [],
    "tagNames": [],
    "taskId": "7001",
    "userGroupIds": [],
    "userId": "6001",
    "workspaceId": "1001",
    "year": 2019,
  },
  {
    "clientId": "3001",
    "description": "This is a test entry (03)",
    "end": 2019-06-24T04:30:00.000Z,
    "entryCount": 0,
    "id": "8003",
    "isActive": false,
    "isBillable": false,
    "isIncluded": false,
    "linkedId": "clock-entry-03",
    "memberOf": "timeEntries",
    "projectId": "2001",
    "start": 2019-06-23T14:30:00.000Z,
    "tagIds": [
      "4003",
    ],
    "tagNames": [
      "tag-c",
    ],
    "taskId": "7002",
    "userGroupIds": [],
    "userId": "6001",
    "workspaceId": "1001",
    "year": 2019,
  },
  {
    "clientId": "3002",
    "description": "This is a test entry (04)",
    "end": 2019-06-23T17:05:20.000Z,
    "entryCount": 0,
    "id": "8004",
    "isActive": false,
    "isBillable": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "timeEntries",
    "projectId": "2002",
    "start": 2019-06-23T14:00:20.000Z,
    "tagIds": [],
    "tagNames": [],
    "taskId": null,
    "userGroupIds": [],
    "userId": "6001",
    "workspaceId": "1001",
    "year": 2019,
  },
  {
    "clientId": "3002",
    "description": "This is a test entry (05)",
    "end": 2019-06-22T19:15:19.000Z,
    "entryCount": 0,
    "id": "8005",
    "isActive": false,
    "isBillable": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "timeEntries",
    "projectId": "2002",
    "start": 2019-06-22T14:00:19.000Z,
    "tagIds": [],
    "tagNames": [],
    "taskId": null,
    "userGroupIds": [],
    "userId": "6001",
    "workspaceId": "1001",
    "year": 2019,
  },
]
`;

exports[`within timeEntriesSelectors the selectors match their snapshots for the sourceTimeEntryCountByTagIdSelector 1`] = `
{
  "4001": 1,
  "4002": 1,
  "4003": 1,
}
`;

exports[`within timeEntriesSelectors the selectors match their snapshots for the timeEntriesTotalCountsByTypeSelector 1`] = `
{
  "existsInTarget": 3,
  "isIncluded": 2,
}
`;

exports[`within timeEntriesSelectors the sourceTimeEntryCountByIdFieldSelectorFactory matches its snapshot 1`] = `{}`;

exports[`within timeEntriesSelectors the timeEntriesForInclusionsTableSelector matches its snapshot based on state.allEntities.areExistsInTargetShown when state.allEntities.areExistsInTargetShown = false 1`] = `
[
  {
    "clientId": "3002",
    "description": "This is a test entry (04)",
    "end": 2019-06-23T17:05:20.000Z,
    "entryCount": 0,
    "existsInTarget": false,
    "id": "8004",
    "isActive": false,
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isBillable": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "timeEntries",
    "projectId": "2002",
    "projectName": "Test Project B",
    "start": 2019-06-23T14:00:20.000Z,
    "tagIds": [],
    "tagNames": [],
    "taskId": null,
    "taskName": null,
    "userGroupIds": [],
    "userId": "6001",
    "workspaceId": "1001",
    "year": 2019,
  },
  {
    "clientId": "3002",
    "description": "This is a test entry (05)",
    "end": 2019-06-22T19:15:19.000Z,
    "entryCount": 0,
    "existsInTarget": false,
    "id": "8005",
    "isActive": false,
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isBillable": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "timeEntries",
    "projectId": "2002",
    "projectName": "Test Project B",
    "start": 2019-06-22T14:00:19.000Z,
    "tagIds": [],
    "tagNames": [],
    "taskId": null,
    "taskName": null,
    "userGroupIds": [],
    "userId": "6001",
    "workspaceId": "1001",
    "year": 2019,
  },
]
`;

exports[`within timeEntriesSelectors the timeEntriesForInclusionsTableSelector matches its snapshot based on state.allEntities.areExistsInTargetShown when state.allEntities.areExistsInTargetShown = true 1`] = `
[
  {
    "clientId": "3001",
    "description": "This is a test entry (01)",
    "end": 2019-06-25T23:00:00.000Z,
    "entryCount": 0,
    "existsInTarget": true,
    "id": "8001",
    "isActive": false,
    "isActiveInSource": true,
    "isActiveInTarget": true,
    "isBillable": false,
    "isIncluded": false,
    "linkedId": "clock-entry-01",
    "memberOf": "timeEntries",
    "projectId": "2001",
    "projectName": "Test Project A",
    "start": 2019-06-25T18:00:00.000Z,
    "tagIds": [
      "4001",
      "4002",
    ],
    "tagNames": [
      "tag-a",
      "tag-b",
    ],
    "taskId": "7001",
    "taskName": "Task 1",
    "userGroupIds": [],
    "userId": "6001",
    "workspaceId": "1001",
    "year": 2019,
  },
  {
    "clientId": "3001",
    "description": "This is a test entry (02)",
    "end": 2019-06-24T22:00:00.000Z,
    "entryCount": 0,
    "existsInTarget": true,
    "id": "8002",
    "isActive": false,
    "isActiveInSource": true,
    "isActiveInTarget": true,
    "isBillable": false,
    "isIncluded": false,
    "linkedId": "clock-entry-02",
    "memberOf": "timeEntries",
    "projectId": "2001",
    "projectName": "Test Project A",
    "start": 2019-06-24T14:00:00.000Z,
    "tagIds": [],
    "tagNames": [],
    "taskId": "7001",
    "taskName": "Task 1",
    "userGroupIds": [],
    "userId": "6001",
    "workspaceId": "1001",
    "year": 2019,
  },
  {
    "clientId": "3001",
    "description": "This is a test entry (03)",
    "end": 2019-06-24T04:30:00.000Z,
    "entryCount": 0,
    "existsInTarget": true,
    "id": "8003",
    "isActive": false,
    "isActiveInSource": true,
    "isActiveInTarget": true,
    "isBillable": false,
    "isIncluded": false,
    "linkedId": "clock-entry-03",
    "memberOf": "timeEntries",
    "projectId": "2001",
    "projectName": "Test Project A",
    "start": 2019-06-23T14:30:00.000Z,
    "tagIds": [
      "4003",
    ],
    "tagNames": [
      "tag-c",
    ],
    "taskId": "7002",
    "taskName": "Task 2",
    "userGroupIds": [],
    "userId": "6001",
    "workspaceId": "1001",
    "year": 2019,
  },
  {
    "clientId": "3002",
    "description": "This is a test entry (04)",
    "end": 2019-06-23T17:05:20.000Z,
    "entryCount": 0,
    "existsInTarget": false,
    "id": "8004",
    "isActive": false,
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isBillable": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "timeEntries",
    "projectId": "2002",
    "projectName": "Test Project B",
    "start": 2019-06-23T14:00:20.000Z,
    "tagIds": [],
    "tagNames": [],
    "taskId": null,
    "taskName": null,
    "userGroupIds": [],
    "userId": "6001",
    "workspaceId": "1001",
    "year": 2019,
  },
  {
    "clientId": "3002",
    "description": "This is a test entry (05)",
    "end": 2019-06-22T19:15:19.000Z,
    "entryCount": 0,
    "existsInTarget": false,
    "id": "8005",
    "isActive": false,
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isBillable": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "timeEntries",
    "projectId": "2002",
    "projectName": "Test Project B",
    "start": 2019-06-22T14:00:19.000Z,
    "tagIds": [],
    "tagNames": [],
    "taskId": null,
    "taskName": null,
    "userGroupIds": [],
    "userId": "6001",
    "workspaceId": "1001",
    "year": 2019,
  },
]
`;

exports[`within timeEntriesSelectors the timeEntriesForInclusionsTableSelector matches its snapshot when state.isDuplicateCheckEnabled = false 1`] = `
[
  {
    "clientId": "3001",
    "description": "This is a test entry (01)",
    "end": 2019-06-25T23:00:00.000Z,
    "entryCount": 0,
    "existsInTarget": false,
    "id": "8001",
    "isActive": false,
    "isActiveInSource": true,
    "isActiveInTarget": true,
    "isBillable": false,
    "isIncluded": false,
    "linkedId": "clock-entry-01",
    "memberOf": "timeEntries",
    "projectId": "2001",
    "projectName": "Test Project A",
    "start": 2019-06-25T18:00:00.000Z,
    "tagIds": [
      "4001",
      "4002",
    ],
    "tagNames": [
      "tag-a",
      "tag-b",
    ],
    "taskId": "7001",
    "taskName": "Task 1",
    "userGroupIds": [],
    "userId": "6001",
    "workspaceId": "1001",
    "year": 2019,
  },
  {
    "clientId": "3001",
    "description": "This is a test entry (02)",
    "end": 2019-06-24T22:00:00.000Z,
    "entryCount": 0,
    "existsInTarget": false,
    "id": "8002",
    "isActive": false,
    "isActiveInSource": true,
    "isActiveInTarget": true,
    "isBillable": false,
    "isIncluded": false,
    "linkedId": "clock-entry-02",
    "memberOf": "timeEntries",
    "projectId": "2001",
    "projectName": "Test Project A",
    "start": 2019-06-24T14:00:00.000Z,
    "tagIds": [],
    "tagNames": [],
    "taskId": "7001",
    "taskName": "Task 1",
    "userGroupIds": [],
    "userId": "6001",
    "workspaceId": "1001",
    "year": 2019,
  },
  {
    "clientId": "3001",
    "description": "This is a test entry (03)",
    "end": 2019-06-24T04:30:00.000Z,
    "entryCount": 0,
    "existsInTarget": false,
    "id": "8003",
    "isActive": false,
    "isActiveInSource": true,
    "isActiveInTarget": true,
    "isBillable": false,
    "isIncluded": false,
    "linkedId": "clock-entry-03",
    "memberOf": "timeEntries",
    "projectId": "2001",
    "projectName": "Test Project A",
    "start": 2019-06-23T14:30:00.000Z,
    "tagIds": [
      "4003",
    ],
    "tagNames": [
      "tag-c",
    ],
    "taskId": "7002",
    "taskName": "Task 2",
    "userGroupIds": [],
    "userId": "6001",
    "workspaceId": "1001",
    "year": 2019,
  },
  {
    "clientId": "3002",
    "description": "This is a test entry (04)",
    "end": 2019-06-23T17:05:20.000Z,
    "entryCount": 0,
    "existsInTarget": false,
    "id": "8004",
    "isActive": false,
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isBillable": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "timeEntries",
    "projectId": "2002",
    "projectName": "Test Project B",
    "start": 2019-06-23T14:00:20.000Z,
    "tagIds": [],
    "tagNames": [],
    "taskId": null,
    "taskName": null,
    "userGroupIds": [],
    "userId": "6001",
    "workspaceId": "1001",
    "year": 2019,
  },
  {
    "clientId": "3002",
    "description": "This is a test entry (05)",
    "end": 2019-06-22T19:15:19.000Z,
    "entryCount": 0,
    "existsInTarget": false,
    "id": "8005",
    "isActive": false,
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isBillable": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "timeEntries",
    "projectId": "2002",
    "projectName": "Test Project B",
    "start": 2019-06-22T14:00:19.000Z,
    "tagIds": [],
    "tagNames": [],
    "taskId": null,
    "taskName": null,
    "userGroupIds": [],
    "userId": "6001",
    "workspaceId": "1001",
    "year": 2019,
  },
]
`;
